### PR TITLE
Added support for AssertionConsumerServiceIndex

### PR DESCRIPTION
**Pull request by joshpfosi**
## Status
**READY**

## Migrations
NO

## Description
This pull request adds support for the `AssertionConsumerServiceIndex` to allow
for multiple development environments to work with an inflexible IdP.

## Todos
- I am unable to test this in production as I am using `ruby-saml` in a gem
which incorporate into many other apps. As Rails does not allow gem dependencies
to be local paths or Git/Github repositories, I need this to be merged to fully
test it. This is a very simple PR, however, and so its correctness can likely be
ascertained by inspection.

## Deploy Notes
N/A

## Impacted Areas in Application
List general components of the application that this PR will affect:

* Orthogonal to current features.

### CODE REVIEW

**Comment by joshpfosi** [comment]
This addresses #327.

**Comment by pitbulk** [comment]
What the point of added index support if only 1 ACS URL is described on the settings? I'm lost.

BTW. In order to test ruby-saml on [ruby-saml-example](https://github.com/onelogin/ruby-saml-example) I used to define at the Gemfile
```
gem 'ruby-saml', :path => '/home/projects/ruby-saml
```
and worked.

**Comment by joshpfosi** [comment]
Right, using the path works in a given project. What does not work is using the path in a gem which is then used in that given project. I actually worked around this by building the gem and deploying it onto our gem server.

I am very new to SAML so perhaps I have misunderstood. My IT administrator made it seem as if this was common practice. Namely, we have multiple environments in which we test our application e.g. http://{app-dev, app-staging, app-beta, app}.net/ if you will. Rather than generate and maintain metadata for each of those apps (our IT department was unwilling to do this as they have hundreds of apps w/ SSO), they suggested we use the same entity ID, generate multiple metadata sets each with unique ACS URL and Index. The index (apparently) will be used by the IdP to choose which URL to use.

To summarize: Each of our environments would submit its own "version" of the metadata with an ACS URL and Index. The IdP, however, would have a merged set of metadata with a single map between index and URL.

**Comment by pitbulk** [comment]
Yes, and this is what we discussed on #327.

In the SAML standard you are able to define multiple ACS URLS, but it used to work to define different binding endpoints, for example, this can be found at a SP metadata:

```
 <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-POST" Location="https://sp.example.com/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp" index="0"/>
    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:1.0:profiles:browser-post" Location="https://sp.example.com/simplesaml/module.php/saml/sp/saml1-acs.php/default-sp" index="1"/>
    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:2.0:bindings:HTTP-Artifact" Location="https://sp.example.com/simplesaml/module.php/saml/sp/saml2-acs.php/default-sp" index="2"/>
    <md:AssertionConsumerService Binding="urn:oasis:names:tc:SAML:1.0:profiles:artifact-01" Location="https://sp.example.com/simplesaml/module.php/saml/sp/saml1-acs.php/default-sp/artifact" index="3"/>
```

But this toolkit only uses for the AssertionConsumerService endpoint the HTTP-POST binding, so at the ruby settings we only let to define one unique ACS URL and we assume that its index is 1.

I understand the "trick" that your IT department suggested to you, but as I already told @joshpfosi, support that will add complexity to the current gem:
- Be able to define on the settings more than 1 ACS URL and keep backward compatibility
- Extend login method to specify the ACS_UR/ACS_index to be used on the AuthNRquest
- Extend AuthNRquest constructor to support the ACS_UR/ACS_index desired and build the XML with the right AssertionConsumerServiceURL
- Modify the SAMLResponse validator to cares about multiple ACS_URLS

As you can see multiple changes and complexity for covering your specific case.
That is why I suggested at #327 to extend the classes and methods for your specific scenario.

**Comment by joshpfosi** [comment]
Ahh I see. Thank you for clarifying that, I understand it much better now. Apologies for the inconvenience.
